rpcs3/Emu/Cell/Modules/cellSearch.cpp

```diff
@@ -98,16 +185,55 @@ error_code cellSearchStartListSearch(CellSearchListSearchType type, CellSearchSo
 		return CELL_SEARCH_ERROR_PARAM;
 	}
 
+	switch (type)
+	{
+	case CELL_SEARCH_LISTSEARCHTYPE_MUSIC_ALBUM:
+	case CELL_SEARCH_LISTSEARCHTYPE_MUSIC_GENRE:
+	case CELL_SEARCH_LISTSEARCHTYPE_MUSIC_ARTIST:
+	case CELL_SEARCH_LISTSEARCHTYPE_PHOTO_YEAR:
+	case CELL_SEARCH_LISTSEARCHTYPE_PHOTO_MONTH:
+	case CELL_SEARCH_LISTSEARCHTYPE_PHOTO_ALBUM:
+	case CELL_SEARCH_LISTSEARCHTYPE_PHOTO_PLAYLIST:
+	case CELL_SEARCH_LISTSEARCHTYPE_VIDEO_ALBUM:
+	case CELL_SEARCH_LISTSEARCHTYPE_MUSIC_PLAYLIST:
+		break;
+	default:
+		return CELL_SEARCH_ERROR_PARAM;
+	}
+
+	if (sortOrder != CELL_SEARCH_SORTORDER_ASCENDING && sortOrder != CELL_SEARCH_SORTORDER_DESCENDING)
+	{
+		return CELL_SEARCH_ERROR_PARAM;
+	}
+
 	const auto search = g_fxo->get<search_info>();
 
+	switch (search->state.load())
+	{
+	case SearchState::idle:
+		break;
+	case SearchState::not_initialized:
+	case SearchState::initializing:
+		return CELL_SEARCH_ERROR_NOT_INITIALIZED;
+	case SearchState::finalizing:
+		return CELL_SEARCH_ERROR_FINALIZING;
+	case SearchState::in_progress:
+	case SearchState::canceling:
+		return CELL_SEARCH_ERROR_BUSY;
+	default:
+		return CELL_SEARCH_ERROR_GENERIC;
+	}
+
 	*outSearchId = idm::make<search_object_t>();
 
+	search->state.store(SearchState::in_progress);
 	sysutil_register_cb([=](ppu_thread& ppu) -> s32
 	{
 		vm::var<CellSearchResultParam> resultParam;
 		resultParam->searchId = *outSearchId;
 		resultParam->resultNum = 0; // TODO
 
+		search->state.store(SearchState::idle);
```

Comment:
release

Reply:
also.. it makes sense that the state store should be done after the callback to avoid recursion.
but maybe it should be he tested after the pr.

Reply:
I doubt it was tested, and your comment makes sense.